crates/pera-core/src/dwallet_mpc/mpc_manager.rs

```diff
@@ -0,0 +1,483 @@
+                warn!(
+                    "received a message for an MPC session ID: `{:?}` which does not exist",
+                    session_id
+                );
+                self.malicious_actors.insert(authority_name);
+                return Ok(());
+            }
+        };
+        match session.handle_message(&DWalletMPCMessage {
+            message: message.to_vec(),
+            authority: authority_name,
+        }) {
+            Err(DwalletMPCError::MaliciousParties(malicious_parties)) => {
+                self.flag_parties_as_malicious(&malicious_parties)?;
+                Ok(())
+            }
+            other => other,
+        }
+    }
+
+    /// Convert the indices of the malicious parties to their addresses and store them
+    /// in the malicious actors set.
+    /// New messages from these parties will be ignored.
+    /// todo(zeev): clarify if it's restarted on epoch change.
+    fn flag_parties_as_malicious(&mut self, malicious_parties: &[PartyID]) -> DwalletMPCResult<()> {
+        let malicious_parties_names = malicious_parties
+            .iter()
+            .map(|party_id| {
+                self.epoch_store()?
+                    .committee()
+                    .authority_by_index(*party_id as u32)
+                    .cloned()
+                    .ok_or(DwalletMPCError::AuthorityIndexNotFound(*party_id))
+            })
+            .collect::<DwalletMPCResult<Vec<AuthorityName>>>()?;
+        warn!(
+            "[dWallet MPC] Flagged the following parties as malicious: {:?}",
+            malicious_parties_names
+        );
+        self.malicious_actors.extend(malicious_parties_names);
+        Ok(())
+    }
+
+    /// Spawns a new MPC session if the number of active sessions is below the limit.
+    /// Otherwise, add the session to the pending queue.
+    pub(crate) fn push_new_mpc_session(
+        &mut self,
+        auxiliary_input: Vec<u8>,
+        party: MPCParty,
+        session_info: SessionInfo,
+    ) -> DwalletMPCResult<()> {
+        if self.mpc_sessions.contains_key(&session_info.session_id) {
+            // This should never happen, as the session ID is a Move UniqueID.
+            error!(
+                "received start flow event for session ID {:?} that already exists",
+                &session_info.session_id
+            );
+            return Ok(());
+        }
+        info!(
+            "Received start MPC flow event for session ID {:?}",
+            session_info.session_id
+        );
+        let mut new_session = DWalletMPCSession::new(
+            self.epoch_store.clone(),
+            self.epoch_id,
+            party,
+            MPCSessionStatus::Pending,
+            auxiliary_input,
+            session_info.clone(),
+        );
+        // TODO (#311): Make sure validator don't mark other validators
+        // TODO (#311): as malicious or take any active action while syncing
+        if self.active_sessions_counter > self.max_active_mpc_sessions {
```

Comment:
```suggestion
        if self.active_sessions_counter >= self.max_active_mpc_sessions {
```
I didn't check implications, but there's a possibility this could lead to a bug (Imagine: you make it so that the active sessions counter is out of sync somehow with the boundaries.)

It is an easy fix, but I'm not a machine and I was close to miss this one. Please, make sure whenever you are doing arithmetics, dereferencings, and comparisons: *think of edge cases*.